### PR TITLE
Restricting the maximum size of memo

### PR DESCRIPTION
**Pull request by confuseSUN**
* **The major changes of this PR**

* **The major impacts of this PR**
  - [ ] Impact WASM?
  - [ ] Impact mainnet data compatibility?

* **Extra documentations**

### CODE REVIEW

**Comment by confuseSUN** [comment]
The number of attributes in TracerMemo is uncertain, causing the upper limit of traceMemo size to be uncertain.

**Comment by findora-sun** [comment]
> The number of attributes in TracerMemo is uncertain, causing the upper limit of traceMemo size to be uncertain.

TracerMemo in platform not work yet (Maybe give up this feature ?), just limit it to 0.

File: api/src/xfr/structs.rs [hunk 1]
```diff
@@ -577,6 +595,56 @@ impl OwnerMemo {
     }
 }
 
+/// Check memo size.
+pub fn check_memo_size(output: &BlindAssetRecord, memo: &Option<OwnerMemo>) -> Result<()> {
```
Comment (findora-sun) [inline api/src/xfr/structs.rs:599]: here, we can use `memo.key_type` to check it is ed25519 or secp256k1, no need `output`.

File: api/src/xfr/structs.rs [hunk 2]
```diff
@@ -577,6 +595,56 @@ impl OwnerMemo {
...
+    if !output.amount.is_confidential() && !output.asset_type.is_confidential() {
+        if memo.is_some() {
+            return Err(eg!(NoahError::AXfrVerifierParamsError));
+        }
+        return Ok(());
+    }
+
+    let memo = memo.clone().unwrap(); //safety unwrap
+
+    match output.public_key.inner() {
```
Comment (findora-sun) [inline api/src/xfr/structs.rs:609]: here we can use `memo.key_type` to check it is ed25519 or secp256k1

File: api/src/xfr/structs.rs [hunk 3]
```diff
@@ -577,6 +595,56 @@ impl OwnerMemo {
...
+/// Check memo size.
+pub fn check_memo_size(output: &BlindAssetRecord, memo: &Option<OwnerMemo>) -> Result<()> {
+    if !output.amount.is_confidential() && !output.asset_type.is_confidential() {
+        if memo.is_some() {
+            return Err(eg!(NoahError::AXfrVerifierParamsError));
+        }
+        return Ok(());
+    }
+
+    let memo = memo.clone().unwrap(); //safety unwrap
```
Comment (findora-sun) [inline api/src/xfr/structs.rs:607]: `clone` maybe not required, if just ref, you can use `memo.as_ref().unwrap()`

**Review by weikengchen — APPROVED** [review]

**Pull request merged**